File: 707.design-linked-list.java
/*
 * @lc app=leetcode id=707 lang=java
 *
 * [707] Design Linked List
 */

// @lc code=start
class MyLinkedList {

    static class Node {
        int val;
        Node prev;
        Node next;

        public Node(int val, Node prev, Node next) {
            this.val = val;
            this.prev = prev;
            this.next = next;
        }
    }

    private Node sentinel;
    private int size;

    public MyLinkedList() {
        this.sentinel = new Node(-1, null, null);
        this.sentinel.prev = this.sentinel;
        this.sentinel.next = this.sentinel;
        this.size = 0;
    }

    private Node find(int index) {
        if (index < 0 || index >= this.size) {
            return null;
        }
        Node iter = this.sentinel;
        while (index != 0) {
            iter = iter.next;
            index -= 1;
        }
        return iter;
    }

    public int get(int index) {
        Node iter = find(index);
        if (iter == null) {
            return -1;
        }
        return iter.next.val;
    }

    public void addAtHead(int val) {
        Node node = new Node(val, this.sentinel, this.sentinel.next);
        this.sentinel.next.prev = node;
        this.sentinel.next = node;
        this.size += 1;
    }

    public void addAtTail(int val) {
        Node node = new Node(val, this.sentinel.prev, this.sentinel);
        this.sentinel.prev.next = node;
        this.sentinel.prev = node;
        this.size += 1;
    }

    public void addAtIndex(int index, int val) {
        if (index == this.size) {
            this.addAtTail(val);
            return;
        }
        Node node = find(index);
        if (node == null) {
            return;
        }
        Node n = new Node(val, node, node.next);
        node.next.prev = n;
        node.next = n;
        this.size += 1;
    }

    public void deleteAtIndex(int index) {
        Node node = find(index);
        if (node == null) {
            return;
        }
        if (this.size == 0) {
            return;
        }
        node.next = node.next.next;
        node.next.prev = node;
        this.size -= 1;
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */
// @lc code=end

File: 707.design-linked-list.cpp
/*
 * @lc app=leetcode id=707 lang=cpp
 *
 * [707] Design Linked List
 */

// @lc code=start

#include <memory>

class MyLinkedList {
private:
    struct LinkedListNode {
        int val;
        LinkedListNode *prev;
        LinkedListNode *next;
        LinkedListNode() = default;
        LinkedListNode(int v, LinkedListNode *p, LinkedListNode *n):
            val(v), prev(p), next(n) {}
    };

    std::unique_ptr<LinkedListNode> sentinel;
    size_t sz = 0;

public:
    MyLinkedList() {
        this->sentinel = std::make_unique<LinkedListNode>();
        this->sentinel->val = -1;
        this->sentinel->prev = this->sentinel.get();
        this->sentinel->next = this->sentinel.get();
    }

    int get(int index) {
        if (index < 0 || index >= this->sz) {
            return -1;
        }
        LinkedListNode *iter = this->sentinel->next;
        int cnt = 0;
        while (cnt != index) {
            iter = iter->next;
            cnt += 1;
        }
        return iter->val;
    }

    void addAtHead(int val) {
        LinkedListNode *node = new LinkedListNode(
            val,
            this->sentinel.get(),
            this->sentinel->next
        );
        this->sentinel->next->prev = node;
        this->sentinel->next = node;
        this->sz += 1;
    }

    void addAtTail(int val) {
        LinkedListNode *node = new LinkedListNode(
            val,
            this->sentinel->prev,
            this->sentinel.get()
        );
        this->sentinel->prev->next = node;
        this->sentinel->prev = node;
        this->sz += 1;
    }

    void addAtIndex(int index, int val) {
        if (index < 0 || index > this->sz) {
            return;
        }
        LinkedListNode *iter = this->sentinel.get();
        int cnt = 0;
        while (cnt != index) {
            iter = iter->next;
            cnt += 1;
        }
        LinkedListNode *node = new LinkedListNode(
            val,
            iter,
            iter->next
        );
        iter->next->prev = node;
        iter->next = node;
        this->sz += 1;
    }

    void deleteAtIndex(int index) {
        if (index < 0 || index >= this->sz) {
            return;
        }
        LinkedListNode *iter = this->sentinel.get();
        int cnt = 0;
        while (cnt != index) {
            iter = iter->next;
            cnt += 1;
        }
        LinkedListNode *tmp = iter->next;
        iter->next = tmp->next;
        tmp->next->prev = iter;
        delete tmp;
        this->sz -= 1;
    }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */
// @lc code=end
